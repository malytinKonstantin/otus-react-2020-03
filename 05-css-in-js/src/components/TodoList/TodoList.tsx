import React, { useState } from 'react'
import {
  TodoListItem as TodoListItemComponent,
  TodoListCreateItem,
} from './components'
import type {
  TodoListItem,
  TodoListHandlers,
  TodoListCreateItemProps,
} from './components'

interface TodoListProps extends TodoListHandlers, TodoListCreateItemProps {
  list: TodoListItem[]
}

export const TodoList: React.FC<TodoListProps> = ({
  list,
  onCreate,
  onActive,
  onDone,
  onRemove,
}) => {
  return (
    <>
      <TodoListCreateItem onCreate={onCreate} />
      {list.map((listItem: TodoListItem) => {
        const { id } = listItem
        return (
          <TodoListItemComponent
            key={id}
            {...listItem}
            onActive={() => onActive(id)}
            onDone={() => onDone(id)}
            onRemove={() => onRemove(id)}
          />
        )
      })}
    </>
  )
}
