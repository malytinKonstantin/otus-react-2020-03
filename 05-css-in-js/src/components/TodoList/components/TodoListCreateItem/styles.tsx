import styled from '@emotion/styled'
import { css } from '@emotion/core'

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 15px;

  & * {
    margin-right: 15px;
  }
`
export const TextField = styled.input`
  padding: 5px 10px;
  border-radius: 5px;
  box-shadow: none;
  border: 1px solid #afafaf;
`
export const Button = styled.button`
  border: none;
  color: #fff;
  background: #1d96ff;
  padding: 5px 15px;
  border-radius: 3px;
`

export const ButtonReset = styled.button`
  ${Button} {
  }
`

export const ButtonSubmit = styled.button`
  ${Button} {
  }
`
