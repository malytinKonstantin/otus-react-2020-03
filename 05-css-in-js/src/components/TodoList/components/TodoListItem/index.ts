export * from './TodoListItem'
export * from './styles'
