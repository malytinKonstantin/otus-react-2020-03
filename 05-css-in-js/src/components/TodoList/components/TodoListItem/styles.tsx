import styled from '@emotion/styled'
import { css } from '@emotion/core'

export const Wrapper = styled.div`
  margin-bottom: 10px;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
`

export const Title = styled.div`
  font-family: monospace;
  font-size: 16px;
  color: #000;
  display: flex;
  cursor: pointer;
`

export const Checkbox = styled.div`
  width: 16px;
  flex-basis: 6px;
  min-width: 16px;
  height: 16px;
  border: 1px solid blue;
  border-radius: 3px;
  margin-right: 10px;
  position: relative;

  &:before {
    content: '';
    position: absolute;
    top: 3px;
    left: 3px;
    width: 10px;
    height: 10px;
    border-radius: 2px;
    background-color: ${({ isCompleted }) =>
      isCompleted ? 'blue' : 'transparent'};
  }

  &:hover {
    cursor: pointer;
  }
`

const ButtonBase = css`
  border: none;
  color: #fff;
  padding: 3px 10px;
  border-radius: 3px;

  &:active {
    outline: none;
    box-shadow: none;
  }
`

export const Button = styled.button`
  ${ButtonBase}
  background-color: #0eb902;
`

export const ButtonRemove = styled.button`
  ${ButtonBase}
  background-color: #ec2c2c;
  margin-left: 10px;
`
