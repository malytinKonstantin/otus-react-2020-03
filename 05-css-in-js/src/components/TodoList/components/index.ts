export * from './TodoListItem'
export * from './TodoListCreateItem'
