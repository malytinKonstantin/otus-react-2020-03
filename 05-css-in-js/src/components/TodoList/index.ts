export * from './TodoList'
export * from './TodoListContainer'
export * from './components'
