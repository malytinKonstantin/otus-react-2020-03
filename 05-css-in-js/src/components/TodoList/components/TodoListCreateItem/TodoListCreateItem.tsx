import React, { useState, useCallback } from 'react'
import { Formik, Form, Field } from 'formik'
import { Wrapper, TextField, ButtonReset, ButtonSubmit } from './styles'

import type { TodoListItem } from '../TodoListItem'

export interface TodoListCreateItemProps {
  /*
   * callback from parent for submit created item
   */
  onCreate: (listItem: TodoListItem) => void
}

export const TodoListCreateItem: React.FC<TodoListCreateItemProps> = (
  props,
) => {
  const handleSubmit = (values, form) => {
    const item = {
      id: Math.random(),
      title: values.title,
      isCompleted: false,
    }
    props.onCreate(item)
    form.resetForm()
  }

  return (
    <Formik
      onSubmit={handleSubmit}
      initialValues={{
        title: '',
      }}
    >
      {(props) => (
        <Form>
          <Wrapper>
            <Field
              type="text"
              name="title"
              id="title"
              component={TextField}
              value={props.values.title}
              onChange={props.handleChange}
            />
            <ButtonSubmit type="submit">добавить</ButtonSubmit>
            <ButtonReset type="button" onClick={props.resetForm}>
              сброс
            </ButtonReset>
          </Wrapper>
        </Form>
      )}
    </Formik>
  )
}
