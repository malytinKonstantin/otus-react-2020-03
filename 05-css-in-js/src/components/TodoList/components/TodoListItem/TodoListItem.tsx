import React from 'react'
import { Title, Checkbox, Button, ButtonRemove, Wrapper, Nav } from './styles'

export interface TodoListItem {
  id: string
  title: string
  isCompleted: boolean
}

interface TodoListItemProps extends TodoListItem {
  onDone: () => void
  onActive: () => void
  onRemove: () => void
}

export const TodoListItem: React.FC<Props> = (props) => {
  const { title, isCompleted, onDone, onActive, onRemove } = props

  return (
    <Wrapper>
      <Title onClick={isCompleted ? onActive : onDone}>
        <Checkbox isCompleted={isCompleted} />
        {title}
      </Title>
      <ButtonRemove onClick={onRemove}>удалить</ButtonRemove>
    </Wrapper>
  )
}
