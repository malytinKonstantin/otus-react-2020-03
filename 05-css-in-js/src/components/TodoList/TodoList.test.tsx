import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import { TodoList } from './TodoList'

import {
  TodoListItem,
  TodoListCreateItem,
  ButtonToDone,
  ButtonToActive,
} from './components'
import { TextField } from './components/TodoListCreateItem/styles'
import { ButtonRemove } from './components/TodoListItem/styles'

describe('<TodoList />', () => {
  const todoList = [
    {
      id: 0,
      title: 'task 1',
      descriprion: 'learn frontend',
      isCompleted: false,
    },
  ]

  const wrapper = mount(<TodoList list={todoList} />)

  it('renders snapshot', () => {
    expect(
      renderer.create(<TodoList list={todoList} />).toJSON(),
    ).toMatchSnapshot()
  })

  it('Default render', () => {
    expect(wrapper.find('TodoListCreateItem')).toHaveLength(1)
    expect(wrapper.find('TodoListItem')).toHaveLength(1)
  })

  it('Add task item', () => {
    wrapper
      .find('TodoListCreateItem')
      .find('input')
      .simulate('change', { target: { value: 'test item' } })
    wrapper.find('TodoListCreateItem').find('.btn-add').at(0).simulate('click')
    expect(wrapper.find('TodoListItem')).toHaveLength(2)
  })

  it('Remove task item', () => {
    wrapper
      .find('TodoListItem')
      .at(1)
      .find('.btn-remove')
      .at(0)
      .simulate('click')
    expect(wrapper.find('TodoListItem')).toHaveLength(1)
  })
})
