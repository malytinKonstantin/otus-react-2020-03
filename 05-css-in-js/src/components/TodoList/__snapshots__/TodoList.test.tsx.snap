// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> renders snapshot 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-bottom: 15px;
}

.emotion-6 * {
  margin-right: 15px;
}

.emotion-0 {
  padding: 5px 10px;
  border-radius: 5px;
  box-shadow: none;
  border: 1px solid #afafaf;
}

.emotion-14 {
  margin-bottom: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-10 {
  font-family: monospace;
  font-size: 16px;
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
}

.emotion-8 {
  width: 16px;
  -webkit-flex-basis: 6px;
  -ms-flex-preferred-size: 6px;
  flex-basis: 6px;
  min-width: 16px;
  height: 16px;
  border: 1px solid blue;
  border-radius: 3px;
  margin-right: 10px;
  position: relative;
}

.emotion-8:before {
  content: '';
  position: absolute;
  top: 3px;
  left: 3px;
  width: 10px;
  height: 10px;
  border-radius: 2px;
  background-color: transparent;
}

.emotion-8:hover {
  cursor: pointer;
}

.emotion-12 {
  border: none;
  color: #fff;
  padding: 3px 10px;
  border-radius: 3px;
  background-color: #ec2c2c;
  margin-left: 10px;
}

.emotion-12:active {
  outline: none;
  box-shadow: none;
}

<div>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="emotion-6 emotion-7"
    >
      <input
        className="emotion-0 emotion-1"
        form={
          Object {
            "dirty": false,
            "errors": Object {},
            "getFieldHelpers": [Function],
            "getFieldMeta": [Function],
            "getFieldProps": [Function],
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleReset": [Function],
            "handleSubmit": [Function],
            "initialErrors": Object {},
            "initialStatus": undefined,
            "initialTouched": Object {},
            "initialValues": Object {
              "title": "",
            },
            "isSubmitting": false,
            "isValid": true,
            "isValidating": false,
            "registerField": [Function],
            "resetForm": [Function],
            "setErrors": [Function],
            "setFieldError": [Function],
            "setFieldTouched": [Function],
            "setFieldValue": [Function],
            "setFormikState": [Function],
            "setStatus": [Function],
            "setSubmitting": [Function],
            "setTouched": [Function],
            "setValues": [Function],
            "status": undefined,
            "submitCount": 0,
            "submitForm": [Function],
            "touched": Object {},
            "unregisterField": [Function],
            "validateField": [Function],
            "validateForm": [Function],
            "validateOnBlur": true,
            "validateOnChange": true,
            "validateOnMount": false,
            "values": Object {
              "title": "",
            },
          }
        }
        id="title"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="emotion-2 emotion-3"
        type="submit"
      >
        добавить
      </button>
      <button
        className="emotion-4 emotion-5"
        onClick={[Function]}
        type="button"
      >
        сброс
      </button>
    </div>
  </form>
  <div
    className="emotion-14 emotion-15"
  >
    <div
      className="emotion-10 emotion-11"
      onClick={[Function]}
    >
      <div
        className="emotion-8 emotion-9"
      />
      task 1
    </div>
    <button
      className="emotion-12 emotion-13"
      onClick={[Function]}
    >
      удалить
    </button>
  </div>
</div>
`;
